### ISSUE TITLE
How to achieve both 'block' and 'inline' Markdown highlighting?

### ISSUE THREAD

**Issue by dvclmn** [krzyzanowskim/STTextView-Plugin-Neon#10]
Hi Marcin, thanks for all the incredible TextKit-related work you've done — and shared. Your projects have been an invaluable resource.

## Issue
When assigning `language: .markdown` as the `NeonPlugin` language, only block grammar such as fenced code blocks, quotes, lists, headings etc are styled. 

As I understand it, TreeSitter can/does use a two-pass parse (ha) for markdown, for block and inline grammar? Is this correct? If so, will I need to implement my own code to achieve this second parsing operation — or indeed, is this even possible? Based on my very tentative TextKit intuition, this sounds expensive. But perhaps (hopefully!) I'm just missing a trick.

## Reproduction
I have created an `STTextView` as per the basic defaults, as outlined in the readme's for this and the [main project](https://github.com/krzyzanowskim/STTextView).

```swift
STTextViewUI.TextView(
    text: $text,
    selection: $selection,
    options: [.wrapLines, .highlightSelectedLine],
    plugins: [NeonPlugin(theme: .default, language: .markdown)]
)
```

My hope was to see such strings as "`inline code`", "*italic text*" etc styled in the typical fashion, but only the block elements mentioned above are receiving styling.

Thanks in advance for any help you can provide.

**Comment by jeanetienne** [comment]
Keen on this too, but I have no idea how to do it. Would you have any pointers Marcin?

**Comment by jeanetienne** [comment]
Good point. I'll investigate with `Neon` and/or `tree-sitter` and report back in this thread if I find something to do in `STTextView-Plugin-Neon` too 👍

**Comment by jeanetienne** [comment]
Based on a few hours of investigation and reading code, my understanding is as follows:

- The implementation/wiring will reside in the STPluginNeon plugin. Neon only offers the _platform_ to integrate "token providers" (among which TreeSitter languages are one option), and TreeSitter only offers the standardised grammars for each language.
- The TreeSitter Markdown grammar explicitly mentions doing a 2 pass for markdown block and markdown inline (see penultimate paragraph of README [here](https://github.com/tree-sitter-grammars/tree-sitter-markdown?tab=readme-ov-file#standalone-usage).)
- STPluginNeon already includes the grammars for markdown (block) and markdown inline.
- I noticed the TreeSitter markdown grammar mentions markdown_inline as a "language" that can be injected in markdown (like HTML and YAML can be injected in Markdown). However it seems the current STPluginNeon implementation doesn't leverage the injection queries? 

Could the injection queries be the mechanism to achieve the second pass? By looking at the git history of the tree-sitter-markdown's repo, I have the impression the comment at the bottom of the README may be outdated, and the code comments at the top of their sample implementation (in `bindings/rust/lib.rs`) seems to also be out of date with the code.

I also had a look at Simon B. Støvring's Runestone iOS app. It successfully does Markdown block and inline highlighting. I had a look at his [Runestone](https://github.com/simonbs/Runestone) and [TreeSitterLanguages](https://github.com/simonbs/TreeSitterLanguages) repos but couldn't find evidence of a "manual" two pass. I then followed his documentation to integrate Runestone (the text view) into a sample iOS app, and added syntax highlighting for markdown. The result is that it _doesn't_ highlight inline markdown automatically... 

### Conclusion
I have the feeling that injection alone is not sufficient, and achieving a good block & inline highlighting is probably down to some sort of "special case", or at least some advanced understanding of how TreeSitter works. Knowledge I don't have. 

@krzyzanowskim, does this word salad make sense to you 😅? 
I'd be happy to help write the implementation but the domain knowledge is quite far above my head, I'd need to (initial) guidance. Alternatively, Simon B. Støvring is probably doing this in the closed source part of the iOS app, we could ask him for pointers?

I hope this helps 🙂

**Comment by jeanetienne** [comment]
Sweet, thank you @mattmassicotte for chiming in! 

I played a bit with the sample project you linked and indeed managed to have a sample MD string to work 👍
(I couldn't highlight properly some "emphasis" and "strong emphasis" but I don't know if it's because of my quick & hacky hardcoded token styling, or from the grammar itself? Anyway, that's secondary to the main effort in this thread).

I barely understand the code in the plugin, let alone in Neon or TreeSitter... So I don't feel qualified to propose a PR just yet 😅

@krzyzanowskim let me know what's your recommendation/approach in terms of architecture in the plugin?

**Comment by jeanetienne** [comment]
> I'd be happy to look more closely with you though!

Sure! How do you want to go about it?

**Comment by mattmassicotte** [comment]
@jeanetienne Here I've opened an issue: https://github.com/ChimeHQ/Neon/issues/51
